sdk/program/src/vote/state/vote_state_deserialize.rs

```diff
@@ -0,0 +1,203 @@
+    let mut buf = [0; 1];
+    cursor
+        .read_exact(&mut buf)
+        .map_err(|_| InstructionError::InvalidAccountData)?;
+
+    Ok(buf[0])
+}
+
+pub(super) fn read_u32(cursor: &mut Cursor<&[u8]>) -> Result<u32, InstructionError> {
+    let mut buf = [0; 4];
+    cursor
+        .read_exact(&mut buf)
+        .map_err(|_| InstructionError::InvalidAccountData)?;
+
+    Ok(u32::from_le_bytes(buf))
+}
+
+fn read_u64(cursor: &mut Cursor<&[u8]>) -> Result<u64, InstructionError> {
+    let mut buf = [0; 8];
+    cursor
+        .read_exact(&mut buf)
+        .map_err(|_| InstructionError::InvalidAccountData)?;
+
+    Ok(u64::from_le_bytes(buf))
+}
+
+fn read_maybe_u64(cursor: &mut Cursor<&[u8]>) -> Result<Option<u64>, InstructionError> {
+    let variant = read_u8(cursor)?;
+    match variant {
+        0 => Ok(None),
+        1 => read_u64(cursor).map(Some),
+        _ => Err(InstructionError::InvalidAccountData),
+    }
+}
+
+fn read_i64(cursor: &mut Cursor<&[u8]>) -> Result<i64, InstructionError> {
+    let mut buf = [0; 8];
+    cursor
+        .read_exact(&mut buf)
+        .map_err(|_| InstructionError::InvalidAccountData)?;
+
+    Ok(i64::from_le_bytes(buf))
+}
+
+fn read_pubkey(cursor: &mut Cursor<&[u8]>) -> Result<Pubkey, InstructionError> {
+    let mut buf = [0; 32];
+    cursor
+        .read_exact(&mut buf)
+        .map_err(|_| InstructionError::InvalidAccountData)?;
+
+    Ok(Pubkey::from(buf))
+}
+
+fn read_votes_into(
+    cursor: &mut Cursor<&[u8]>,
+    vote_state: &mut VoteState,
+    has_latency: bool,
+) -> Result<(), InstructionError> {
+    let vote_count = read_u64(cursor)?;
+
+    for _ in 0..vote_count {
+        let latency = if has_latency { read_u8(cursor)? } else { 0 };
+
+        let slot = read_u64(cursor)?;
+        let confirmation_count = read_u32(cursor)?;
+        let lockout = Lockout::new_with_confirmation_count(slot, confirmation_count);
+
+        vote_state.votes.push_back(LandedVote { latency, lockout });
+    }
+
+    Ok(())
+}
+
+fn read_authorized_voters_into(
+    cursor: &mut Cursor<&[u8]>,
+    vote_state: &mut VoteState,
+) -> Result<(), InstructionError> {
+    let authorized_voter_count = read_u64(cursor)?;
+
+    for _ in 0..authorized_voter_count {
+        let epoch = read_u64(cursor)?;
+        let authorized_voter = read_pubkey(cursor)?;
+
+        vote_state.authorized_voters.insert(epoch, authorized_voter);
+    }
+
+    Ok(())
+}
+
+fn read_prior_voters_into(
+    cursor: &mut Cursor<&[u8]>,
+    vote_state: &mut VoteState,
+) -> Result<(), InstructionError> {
+    for _ in 0..MAX_ITEMS {
+        let prior_voter = read_pubkey(cursor)?;
+        let from_epoch = read_u64(cursor)?;
+        let until_epoch = read_u64(cursor)?;
+        let item = (prior_voter, from_epoch, until_epoch);
+
+        // HANA someone from the core team should confirm this holds
+        match item {
+            (_, 0, 0) => (),
+            _ => vote_state.prior_voters.append(item),
+        }
+    }
```

Comment:
in the cool light of morning this is clearly incorrect, if the ringbuffer has wrapped around then we will end up with the wrong idx value. will change this function to write directly into the item buffer and set the index by hand